File: src/main/java/io/github/tamawish/pureeconomy/hook/VaultEconomy.java
package io.github.tamawish.pureeconomy.hook;

import io.github.tamawish.pureeconomy.PureEconomy;
import io.github.tamawish.pureeconomy.economy.Currency;
import io.github.tamawish.pureeconomy.economy.EconomyService;
import net.milkbowl.vault.economy.AbstractEconomy;
import net.milkbowl.vault.economy.EconomyResponse;
import org.bukkit.Bukkit;
import org.bukkit.OfflinePlayer;

import java.math.BigDecimal;
import java.util.Collections;
import java.util.List;
import java.util.UUID;

public final class VaultEconomy extends AbstractEconomy {

    private final PureEconomy plugin;

    public VaultEconomy(PureEconomy plugin) {
        this.plugin = plugin;
    }

    private EconomyService eco() {
        return plugin.economy();
    }

    private Currency cur() {
        return eco().defaultCurrency();
    }

    private UUID id(String name) {
        return eco().resolve(name);
    }

    @Override
    public boolean isEnabled() {
        return plugin.isEnabled() && cur() != null;
    }

    @Override
    public String getName() {
        return "PureEconomy";
    }

    @Override
    public boolean hasBankSupport() {
        return false;
    }

    @Override
    public int fractionalDigits() {
        return cur() == null ? 2 : cur().decimals();
    }

    @Override
    public String format(double amount) {
        if (cur() == null) {
            return String.valueOf(amount);
        }
        return cur().format(BigDecimal.valueOf(amount));
    }

    @Override
    public String currencyNamePlural() {
        return cur() == null ? "Coins" : cur().plural();
    }

    @Override
    public String currencyNameSingular() {
        return cur() == null ? "Coin" : cur().singular();
    }

    @Override
    public boolean hasAccount(String playerName) {
        return id(playerName) != null;
    }

    @Override
    public boolean hasAccount(String playerName, String worldName) {
        return hasAccount(playerName);
    }

    @Override
    public boolean hasAccount(OfflinePlayer player) {
        return player != null;
    }

    @Override
    public boolean hasAccount(OfflinePlayer player, String worldName) {
        return hasAccount(player);
    }

    @Override
    public double getBalance(String playerName) {
        UUID uuid = id(playerName);
        if (uuid == null || cur() == null) {
            return 0;
        }
        return eco().get(uuid, cur()).doubleValue();
    }

    @Override
    public double getBalance(String playerName, String world) {
        return getBalance(playerName);
    }

    @Override
    public double getBalance(OfflinePlayer player) {
        if (player == null || cur() == null) {
            return 0;
        }
        eco().ensureAccount(player.getUniqueId(), player.getName());
        return eco().get(player.getUniqueId(), cur()).doubleValue();
    }

    @Override
    public double getBalance(OfflinePlayer player, String world) {
        return getBalance(player);
    }

    @Override
    public boolean has(String playerName, double amount) {
        UUID uuid = id(playerName);
        if (uuid == null || cur() == null) {
            return false;
        }
        return eco().has(uuid, cur(), BigDecimal.valueOf(amount));
    }

    @Override
    public boolean has(String playerName, String worldName, double amount) {
        return has(playerName, amount);
    }

    @Override
    public boolean has(OfflinePlayer player, double amount) {
        if (player == null || cur() == null) {
            return false;
        }
        return eco().has(player.getUniqueId(), cur(), BigDecimal.valueOf(amount));
    }

    @Override
    public boolean has(OfflinePlayer player, String worldName, double amount) {
        return has(player, amount);
    }

    @Override
    public EconomyResponse withdrawPlayer(String playerName, double amount) {
        UUID uuid = id(playerName);
        if (uuid == null) {
            return new EconomyResponse(0, 0, EconomyResponse.ResponseType.FAILURE, "Unknown player");
        }
        return withdraw(uuid, amount);
    }

    @Override
    public EconomyResponse withdrawPlayer(String playerName, String worldName, double amount) {
        return withdrawPlayer(playerName, amount);
    }

    @Override
    public EconomyResponse withdrawPlayer(OfflinePlayer player, double amount) {
        if (player == null) {
            return new EconomyResponse(0, 0, EconomyResponse.ResponseType.FAILURE, "Unknown player");
        }
        return withdraw(player.getUniqueId(), amount);
    }

    @Override
    public EconomyResponse withdrawPlayer(OfflinePlayer player, String worldName, double amount) {
        return withdrawPlayer(player, amount);
    }

    private EconomyResponse withdraw(UUID uuid, double amount) {
        if (cur() == null) {
            return fail("No default currency");
        }
        if (!Double.isFinite(amount) || amount <= 0) {
            return fail("Amount must be finite and positive");
        }
        BigDecimal bd = cur().normalize(BigDecimal.valueOf(amount));
        if (bd.compareTo(BigDecimal.ZERO) <= 0) {
            return fail("Amount is below the minimum currency precision");
        }
        if (!eco().take(uuid, cur(), bd)) {
            return new EconomyResponse(0, eco().get(uuid, cur()).doubleValue(), EconomyResponse.ResponseType.FAILURE, "Insufficient funds");
        }
        return new EconomyResponse(bd.doubleValue(), eco().get(uuid, cur()).doubleValue(), EconomyResponse.ResponseType.SUCCESS, null);
    }

    @Override
    public EconomyResponse depositPlayer(String playerName, double amount) {
        UUID uuid = id(playerName);
        if (uuid == null) {
            return fail("Unknown player");
        }
        return deposit(uuid, amount);
    }

    @Override
    public EconomyResponse depositPlayer(String playerName, String worldName, double amount) {
        return depositPlayer(playerName, amount);
    }

    @Override
    public EconomyResponse depositPlayer(OfflinePlayer player, double amount) {
        if (player == null) {
            return fail("Unknown player");
        }
        return deposit(player.getUniqueId(), amount);
    }

    @Override
    public EconomyResponse depositPlayer(OfflinePlayer player, String worldName, double amount) {
        return depositPlayer(player, amount);
    }

    private EconomyResponse deposit(UUID uuid, double amount) {
        if (cur() == null) {
            return fail("No default currency");
        }
        if (!Double.isFinite(amount) || amount <= 0) {
            return fail("Amount must be finite and positive");
        }
        BigDecimal bd = cur().normalize(BigDecimal.valueOf(amount));
        if (bd.compareTo(BigDecimal.ZERO) <= 0) {
            return fail("Amount is below the minimum currency precision");
        }
        if (!eco().add(uuid, cur(), bd)) {
            return fail("Max balance");
        }
        return new EconomyResponse(bd.doubleValue(), eco().get(uuid, cur()).doubleValue(), EconomyResponse.ResponseType.SUCCESS, null);
    }

    @Override
    public EconomyResponse createBank(String name, String player) {
        return noBank();
    }

    @Override
    public EconomyResponse createBank(String name, OfflinePlayer player) {
        return noBank();
    }

    @Override
    public EconomyResponse deleteBank(String name) {
        return noBank();
    }

    @Override
    public EconomyResponse bankBalance(String name) {
        return noBank();
    }

    @Override
    public EconomyResponse bankHas(String name, double amount) {
        return noBank();
    }

    @Override
    public EconomyResponse bankWithdraw(String name, double amount) {
        return noBank();
    }

    @Override
    public EconomyResponse bankDeposit(String name, double amount) {
        return noBank();
    }

    @Override
    public EconomyResponse isBankOwner(String name, String playerName) {
        return noBank();
    }

    @Override
    public EconomyResponse isBankOwner(String name, OfflinePlayer player) {
        return noBank();
    }

    @Override
    public EconomyResponse isBankMember(String name, String playerName) {
        return noBank();
    }

    @Override
    public EconomyResponse isBankMember(String name, OfflinePlayer player) {
        return noBank();
    }

    @Override
    public List<String> getBanks() {
        return Collections.emptyList();
    }

    @Override
    public boolean createPlayerAccount(String playerName) {
        UUID uuid = id(playerName);
        if (uuid == null) {
            OfflinePlayer off = Bukkit.getOfflinePlayer(playerName);
            eco().ensureAccount(off.getUniqueId(), playerName);
            return true;
        }
        eco().ensureAccount(uuid, playerName);
        return true;
    }

    @Override
    public boolean createPlayerAccount(String playerName, String worldName) {
        return createPlayerAccount(playerName);
    }

    @Override
    public boolean createPlayerAccount(OfflinePlayer player) {
        if (player == null) {
            return false;
        }
        eco().ensureAccount(player.getUniqueId(), player.getName());
        return true;
    }

    @Override
    public boolean createPlayerAccount(OfflinePlayer player, String worldName) {
        return createPlayerAccount(player);
    }

    private static EconomyResponse noBank() {
        return new EconomyResponse(0, 0, EconomyResponse.ResponseType.NOT_IMPLEMENTED, "No banks");
    }

    private static EconomyResponse fail(String reason) {
        return new EconomyResponse(0, 0, EconomyResponse.ResponseType.FAILURE, reason);
    }
}

File: src/main/kotlin/io/github/tamawish/pureeconomy/hook/VaultEconomy.kt
package io.github.tamawish.pureeconomy.hook

import io.github.tamawish.pureeconomy.PureEconomy
import io.github.tamawish.pureeconomy.economy.Currency
import io.github.tamawish.pureeconomy.economy.EconomyService
import net.milkbowl.vault.economy.AbstractEconomy
import net.milkbowl.vault.economy.EconomyResponse
import org.bukkit.OfflinePlayer
import java.math.BigDecimal
import java.util.UUID

/** Vault [net.milkbowl.vault.economy.Economy] adapter for PureEconomy's default currency. */
class VaultEconomy(
    private val plugin: PureEconomy,
) : AbstractEconomy() {
    private fun eco(): EconomyService = plugin.economy()

    private fun cur(): Currency? = eco().defaultCurrency()

    private fun id(name: String): UUID? = eco().resolve(name, lookupOffline = false)

    override fun isEnabled(): Boolean = plugin.isEnabled && cur() != null

    override fun getName(): String = "PureEconomy"

    override fun hasBankSupport(): Boolean = false

    override fun fractionalDigits(): Int = cur()?.decimals() ?: 2

    override fun format(amount: Double): String {
        val currency = cur() ?: return amount.toString()
        return currency.format(BigDecimal.valueOf(amount))
    }

    override fun currencyNamePlural(): String = cur()?.plural() ?: "Coins"

    override fun currencyNameSingular(): String = cur()?.singular() ?: "Coin"

    override fun hasAccount(playerName: String): Boolean = id(playerName) != null

    override fun hasAccount(
        playerName: String,
        worldName: String,
    ): Boolean = hasAccount(playerName)

    override fun hasAccount(player: OfflinePlayer?): Boolean = player != null

    override fun hasAccount(
        player: OfflinePlayer?,
        worldName: String,
    ): Boolean = hasAccount(player)

    override fun getBalance(playerName: String): Double {
        val uuid = id(playerName)
        val currency = cur()
        if (uuid == null || currency == null) {
            return 0.0
        }
        return eco().get(uuid, currency).toDouble()
    }

    override fun getBalance(
        playerName: String,
        world: String,
    ): Double = getBalance(playerName)

    override fun getBalance(player: OfflinePlayer?): Double {
        val currency = cur()
        if (player == null || currency == null) {
            return 0.0
        }
        eco().ensureAccount(player.uniqueId, player.name)
        return eco().get(player.uniqueId, currency).toDouble()
    }

    override fun getBalance(
        player: OfflinePlayer?,
        world: String,
    ): Double = getBalance(player)

    override fun has(
        playerName: String,
        amount: Double,
    ): Boolean {
        val uuid = id(playerName)
        val currency = cur()
        if (uuid == null || currency == null) {
            return false
        }
        return eco().has(uuid, currency, BigDecimal.valueOf(amount))
    }

    override fun has(
        playerName: String,
        worldName: String,
        amount: Double,
    ): Boolean = has(playerName, amount)

    override fun has(
        player: OfflinePlayer?,
        amount: Double,
    ): Boolean {
        val currency = cur()
        if (player == null || currency == null) {
            return false
        }
        return eco().has(player.uniqueId, currency, BigDecimal.valueOf(amount))
    }

    override fun has(
        player: OfflinePlayer?,
        worldName: String,
        amount: Double,
    ): Boolean = has(player, amount)

    override fun withdrawPlayer(
        playerName: String,
        amount: Double,
    ): EconomyResponse {
        val uuid =
            id(playerName)
                ?: return EconomyResponse(0.0, 0.0, EconomyResponse.ResponseType.FAILURE, "Unknown player")
        return withdraw(uuid, amount)
    }

    override fun withdrawPlayer(
        playerName: String,
        worldName: String,
        amount: Double,
    ): EconomyResponse = withdrawPlayer(playerName, amount)

    override fun withdrawPlayer(
        player: OfflinePlayer?,
        amount: Double,
    ): EconomyResponse {
        if (player == null) {
            return EconomyResponse(0.0, 0.0, EconomyResponse.ResponseType.FAILURE, "Unknown player")
        }
        return withdraw(player.uniqueId, amount)
    }

    override fun withdrawPlayer(
        player: OfflinePlayer?,
        worldName: String,
        amount: Double,
    ): EconomyResponse = withdrawPlayer(player, amount)

    private fun withdraw(
        uuid: UUID,
        amount: Double,
    ): EconomyResponse {
        val currency = cur() ?: return fail("No default currency")
        if (!amount.isFinite() || amount <= 0) {
            return fail("Amount must be finite and positive")
        }
        val value = currency.normalize(BigDecimal.valueOf(amount))
        if (value <= BigDecimal.ZERO) {
            return fail("Amount is below the minimum currency precision")
        }
        val result = eco().takeResult(uuid, currency, value, bank = false)
        if (!result.success) {
            return EconomyResponse(
                0.0,
                (result.from ?: eco().get(uuid, currency)).toDouble(),
                EconomyResponse.ResponseType.FAILURE,
                if (result.status == io.github.tamawish.pureeconomy.api.EconomyStatus.INSUFFICIENT) {
                    "Insufficient funds"
                } else {
                    result.status.name
                },
            )
        }
        return EconomyResponse(
            value.toDouble(),
            (result.from ?: value).toDouble(),
            EconomyResponse.ResponseType.SUCCESS,
            null,
        )
    }

    override fun depositPlayer(
        playerName: String,
        amount: Double,
    ): EconomyResponse {
        val uuid = id(playerName) ?: return fail("Unknown player")
        return deposit(uuid, amount)
    }

    override fun depositPlayer(
        playerName: String,
        worldName: String,
        amount: Double,
    ): EconomyResponse = depositPlayer(playerName, amount)

    override fun depositPlayer(
        player: OfflinePlayer?,
        amount: Double,
    ): EconomyResponse {
        if (player == null) {
            return fail("Unknown player")
        }
        return deposit(player.uniqueId, amount)
    }

    override fun depositPlayer(
        player: OfflinePlayer?,
        worldName: String,
        amount: Double,
    ): EconomyResponse = depositPlayer(player, amount)

    private fun deposit(
        uuid: UUID,
        amount: Double,
    ): EconomyResponse {
        val currency = cur() ?: return fail("No default currency")
        if (!amount.isFinite() || amount <= 0) {
            return fail("Amount must be finite and positive")
        }
        val value = currency.normalize(BigDecimal.valueOf(amount))
        if (value <= BigDecimal.ZERO) {
            return fail("Amount is below the minimum currency precision")
        }
        val result = eco().addResult(uuid, currency, value, bank = false)
        if (!result.success) {
            return fail("Max balance")
        }
        return EconomyResponse(
            value.toDouble(),
            (result.from ?: value).toDouble(),
            EconomyResponse.ResponseType.SUCCESS,
            null,
        )
    }

    override fun createBank(
        name: String,
        player: String,
    ): EconomyResponse = noBank()

    override fun createBank(
        name: String,
        player: OfflinePlayer,
    ): EconomyResponse = noBank()

    override fun deleteBank(name: String): EconomyResponse = noBank()

    override fun bankBalance(name: String): EconomyResponse = noBank()

    override fun bankHas(
        name: String,
        amount: Double,
    ): EconomyResponse = noBank()

    override fun bankWithdraw(
        name: String,
        amount: Double,
    ): EconomyResponse = noBank()

    override fun bankDeposit(
        name: String,
        amount: Double,
    ): EconomyResponse = noBank()

    override fun isBankOwner(
        name: String,
        playerName: String,
    ): EconomyResponse = noBank()

    override fun isBankOwner(
        name: String,
        player: OfflinePlayer,
    ): EconomyResponse = noBank()

    override fun isBankMember(
        name: String,
        playerName: String,
    ): EconomyResponse = noBank()

    override fun isBankMember(
        name: String,
        player: OfflinePlayer,
    ): EconomyResponse = noBank()

    override fun getBanks(): List<String> = emptyList()

    override fun createPlayerAccount(playerName: String): Boolean {
        val uuid = id(playerName) ?: return false
        eco().ensureAccount(uuid, playerName)
        return true
    }

    override fun createPlayerAccount(
        playerName: String,
        worldName: String,
    ): Boolean = createPlayerAccount(playerName)

    override fun createPlayerAccount(player: OfflinePlayer?): Boolean {
        if (player == null) {
            return false
        }
        eco().ensureAccount(player.uniqueId, player.name)
        return true
    }

    override fun createPlayerAccount(
        player: OfflinePlayer?,
        worldName: String,
    ): Boolean = createPlayerAccount(player)

    companion object {
        private fun noBank(): EconomyResponse = EconomyResponse(0.0, 0.0, EconomyResponse.ResponseType.NOT_IMPLEMENTED, "No banks")

        private fun fail(reason: String): EconomyResponse = EconomyResponse(0.0, 0.0, EconomyResponse.ResponseType.FAILURE, reason)
    }
}
